CASSANDRA-20328: Migrate sstableloader code to its own tools directory and artifact

tolbertam opened the pull request:
As part of a broader effort to decouple java driver code from the server code, this moves sstableloader to its own tools directory.

As sstableloader is also used as a library (CASSANDRA-10637) also added a new artifact 'cassandra-sstableloader' that will get deployed to maven along with 'cassandra-all'.  While I expect this is likely a niche use case, this will allow users to continue using BulkExport as a library.

Also updates IDE project files and circleci to utilize new sstableloader specific targets.

For CASSANDRA-20328

Tasks:

- [x] Debug and fix failing python dtests around sstableloader (fixed via [apache/cassandra-ccm#787](https://github.com/apache/cassandra-ccm/pull/787))
- [x] Verify eclipse configuration (eclipse functionality currently broken on trunk, but made some changes to verify the changes on this branch would work after CASSANDRA-20497)
- [x] Verify netbeans configuration (loaded up netbeans and verified sstableloader source and tests were considered  source paths, although generally struggled to get netbeans to recognize classes across source paths)
![Screenshot From 2025-03-30 16-53-19](https://github.com/user-attachments/assets/b70d6403-547a-4f17-a7ed-1a7efad9d426)

tolbertam commented on tools/sstableloader/test/unit/org/apache/cassandra/tools/LoaderOptionsTest.java:
```diff
@@ -215,5 +215,40 @@ public void checkPermission(Permission perm, Object context)
         }
     }
+
+    // Copied from OfflineToolUtils
```
Ultimately ended up doing this instead of having sstableloader-test depend on test classes, as the test classes already depend on the sstableloader build, this didn't feel like it was worth doing, so just went for copying the minimal amount of shared code.

tolbertam commented on ide/nbproject/project.xml:
```diff
@@ -35,6 +35,12 @@
                     <location>${project.dir}/tools/fqltool/src</location>
                     <encoding>UTF-8</encoding>
                 </source-folder>
+                <source-folder>
```
TODO: verify this actually works in netbeans.

michaelsembwever replied:
@michaelsembwever todo

tolbertam replied:
Ah, forgot to note here, I did verify this works in netbeans (screenshot in pr description)

tolbertam commented on build.xml:
```diff
@@ -2008,8 +2100,10 @@
   <classpathentry kind="src" output="${test.classes}" path="test/resources" />
   <classpathentry kind="src" path="tools/stress/src"/>
   <classpathentry kind="src" path="tools/fqltool/src"/>
+  <classpathentry kind="src" path="tools/sstableloader/src"/>
```
TODO: Verify this works in eclipse

tolbertam replied:
This seems to work as I can run `SSTableLoaderTest` and all the tests pass, but I had to make eclipse work again as it looks like it's currently broken for a number of reasons (see other comments).

If you think these changes are out of scope let me know and I can do them as a separate jira/PR.

Note that after doing this there were still a few things i needed to do:

1. Add `-DallowAttachSelf=true` to the VM run arguments for any run configuration.
2. Update `SortedTableWriter.AbstractIndexWriter` to be public as otherwise I get this error:

```
Caused by: java.lang.Error: Unresolved compilation problem: 
	The type SortedTableWriter.AbstractIndexWriter is not visible

	at org.apache.cassandra.io.sstable.format.big.BigTableWriter$IndexWriter.<init>(BigTableWriter.java:237)
	at org.apache.cassandra.io.sstable.format.big.BigTableWriter$Builder.openIndexWriter(BigTableWriter.java:405)
	at org.apache.cassandra.io.sstable.format.big.BigTableWriter$Builder.openIndexWriter(BigTableWriter.java:1)
```

Change made:
```diff
diff --git a/src/java/org/apache/cassandra/io/sstable/format/SortedTableWriter.java b/src/java/org/apache/cassandra/io/sstable/format/SortedTableWriter.java
index 4fca9ca181..67ae1c7402 100644
--- a/src/java/org/apache/cassandra/io/sstable/format/SortedTableWriter.java
+++ b/src/java/org/apache/cassandra/io/sstable/format/SortedTableWriter.java
@@ -427,7 +427,7 @@ public abstract class SortedTableWriter<P extends SortedTablePartitionWriter, I
         }
     }

-    protected static abstract class AbstractIndexWriter extends AbstractTransactional implements Transactional
+    public static abstract class AbstractIndexWriter extends AbstractTransactional implements Transactional
     {
         protected final Descriptor descriptor;
         protected final TableMetadataRef metadata;
```

I suspect this isn't actually necessary, I think maybe what is happening is the code ends up somewhere else on the build path, as I see duplication in the jars being added, so the same may be happening on the classpath.  I'm not sure how much further I should to take this other than to ensure I can at a base level get things working.

tolbertam replied:
Can ignore this commentary (doing the unrelated eclipse changes in #4020 instead)

tolbertam commented on build.xml:
```diff
@@ -2064,6 +2158,16 @@
              classifier="javadoc"/>
+
+    <!-- the cassandra-sstableloader jar -->
+    <install pomFile="${build.dir}/${final.name}-sstableloader.pom"
```
Tested this locally and it does appear to work:

```kotlin
plugins {
    id("java")
}

group = "org.example"
version = "1.0-SNAPSHOT"

repositories {
    mavenCentral()
    mavenLocal()
}

dependencies {
    implementation("org.apache.cassandra:cassandra-sstableloader:5.1-SNAPSHOT")
    testImplementation(platform("org.junit:junit-bom:5.10.0"))
    testImplementation("org.junit.jupiter:junit-jupiter")
}

tasks.test {
    useJUnitPlatform()
}
```

![Screenshot From 2025-02-13 19-37-11](https://github.com/user-attachments/assets/ad767c1b-4670-4d63-a215-9654fda4269a)

tolbertam commented on build.xml:
```diff
@@ -64,6 +64,7 @@
     <property name="build.classes" value="${build.dir}/classes"/>
     <property name="build.classes.main" value="${build.classes}/main" />
     <property name="javadoc.dir" value="${build.dir}/javadoc"/>
+    <property name="sstableloader-javadoc.dir" value="${build.dir}/sstableloader-javadoc"/>
```
Move to ~660 where other sstableloader props are, rename to `sstableloader.javadoc.dir` for consistency.

smiklosovic commented on .build/parent-pom-template.xml:
```diff
@@ -671,7 +671,7 @@
       <dependency>
         <groupId>org.apache.cassandra</groupId>
         <artifactId>cassandra-all</artifactId>
-        <version>4.1-alpha2-SNAPSHOT</version>
```
interesting

tolbertam replied:
yeah, this surprised me too!  Ultimately it wasn't being used because I don't think there was a dependency on it (until now).

smiklosovic commented on tools/bin/sstableloader:
```diff
@@ -0,0 +1,49 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+if [ "x$CASSANDRA_INCLUDE" = "x" ]; then
```
@tolbertam please check `redhat/cassandra.spec`. I know that `bin/sstableloader` points to this one, I just wonder what changes, if anything, if we introduce this script among tools.

smiklosovic replied:
tools are in a different package a user has to install so when we move sstableloader among tools, then, suddenly, the script in `bin/sstableloader` will point to a non-existing `tools/bin/sstableloader` (unless a user installs it).

tolbertam replied:
ah, this is not something I had thought about yet, I'll see what I can do about this.

tolbertam replied:
resolving conversation as in other comments noted that was able to make both the debian and redhat builds work :+1:

smiklosovic commented on bin/sstableloader:
```diff
@@ -32,18 +35,6 @@ elif [ -r "$CASSANDRA_INCLUDE" ]; then
-        -Dcassandra.storagedir="$cassandra_storagedir" \
-        -Dlogback.configurationFile=logback-tools.xml \
-        org.apache.cassandra.tools.BulkLoader "$@"
+"$CASSANDRA_HOME/tools/bin/sstableloader" "$@"
```
@tolbertam  I would check the existence of this script because sstableloader is a tool which is / should be installed by a different package (redhat/cassandra.spec).

If it is not present, then we need to tell a user that they should install tools package.

tolbertam replied:
I suppose now that we're bundling sstableloader with tools, using it will now require the cassandra-tools package, where previously it was bundled with cassandra.  I hope this is the kind of change we can make between 5.0 and whatever the next release?

We could in theory still bundle sstableloader with the core package, it's just that it will ultimately its library will live in the tools directory.

I'll go with bundling it in the tools package and updating this scrip to check for presence of the script and erroring out with a message as you suggest.  I will also add something to news for this.

tolbertam replied:
actually, it's interesting;  It looks like both `fqltool` and `cassandra-stress` are packaged in the main cassandra package, not the tools package.  This is a bit contrary to what I expected, but other tools like sstabledump, which don't have their own tools directory/library path are in the tools package.

I think what I should do is just retain that behavior and keep sstableloader packaged in the cassandra package.  I'll just need to test it in a docker container to see that it does the right thing for both redhat and debian.

We'll also need to make sure that the right `sstableloader` binary is copied to `/usr/bin`, which should hopefully be a trivial change.

tolbertam replied:
I tried building both the debian and redhat package and noticed that `sstableloader`, `cassandra-stress` and `fqltool` were all present in the base package, not tools.  Given that is the case, I think its best to continue packaging things this way, and to update the builders to copy the sstableloader jar and the correct binary with it.  Will push a change that does this shortly.

Another thing I noticed was that these jars are always present on the classpath, even for cassandra server.  When I implement [CASSANDRA-20329](https://issues.apache.org/jira/browse/CASSANDRA-20329) I'll make sure that is no longer the case.

smiklosovic replied:
@tolbertam I see. So the plan is to keep packaging sstableloader as is, it would be just a different class path etc. Right?

tolbertam replied:
That's right,  I think given stress and fqltool are bundled in the main distribution package already (even though they have separate builds into tools), we might as well do the same for sstableloader.

smiklosovic commented on bin/sstableloader:
```diff
@@ -32,18 +35,18 @@ elif [ -r "$CASSANDRA_INCLUDE" ]; then
     . "$CASSANDRA_INCLUDE"
 fi
 
-if [ -z "$CLASSPATH" ]; then
```
@tolbertam why have you removed all of this?

```
if [ -z "$CLASSPATH" ]; then
    echo "You must set the CLASSPATH var" >&2
    exit 1
fi

if [ "x$MAX_HEAP_SIZE" = "x" ]; then
    MAX_HEAP_SIZE="256M"
fi
```
edit:

@tolbertam 

Aha I see, because it is all moved under tools/bin/sstableloader. Interesting. That way we might get rid of _everything_ above, no? Just check the script's existence and then call it with all the args. This script should be just a "dumb proxy" (plus checking if the target script exists).

smiklosovic replied:
@tolbertam 

Aha I see, because it is all moved under tools/bin/sstableloader. Interesting. That way we might get rid of _everything_ above, no? Just check the script existence and then call it with all the args.

tolbertam replied:
Yeah, thats a good idea, keeps things simple to make it a dumb proxy, will change.

tolbertam replied:
actually, we still need the existing code because we need to resolve `$CASSANDRA_HOME`.  Alternatively we could just imply the tools directory is in /tools/bin, but knowing that maybe can just keep it?

smiklosovic replied:
ok just keep it like it is

smiklosovic commented on bin/sstableloader:
```diff
@@ -32,18 +35,18 @@ elif [ -r "$CASSANDRA_INCLUDE" ]; then
+issue as this script should not be installed." >&2
+    exit 1
+else
+    echo "Warning: sstableloader has been moved to the tools directory. \
```
@tolbertam I am not sure we want to do this. The logic should be "if the script in tools does not exist, tell user that it moved to tools and they should install that package, otherwise run it".

In other words, there is no reason to emit a warning telling a user that script moved and then execute that script. Why should a user care it moved if it works? I would just remove that "else" branch completely.

smiklosovic commented on bin/sstableloader:
```diff
@@ -32,18 +35,18 @@ elif [ -r "$CASSANDRA_INCLUDE" ]; then
-    MAX_HEAP_SIZE="256M"
+if [ ! -f "$SSTABLELOADER_PATH" ]; then
+    echo "Note: sstableloader has moved to the tools directory. \
+Detected that $SSTABLELOADER_PATH does not exist, this is likely a packaging \
```
I don't get `this is likely a packaging issue as this script should not be installed.`

This script should be still installed. It is just it will point to a new executable. It is just a delegate.

Imho should be `Error: sstableloader has moved to the tools directory. Please install cassandra-tools package and try again.`

smiklosovic replied:
hmmm thinking about it more, I see what we are doing here. This script is still part of tarball, it is just not "installed" as part of deb / rpm. When I look closely, I see this was installed to `/usr/bin/sstableloader` and `tools/bin/sstableloader` can not be installed there as well upon installing tools package because it would replace what was already in `/usr/bin/sstableloader`. Huh.

But still. I don't think this is a _packaging issue_. Because e.g. a developer might still execute this script as he develops. I would just say that it moved to tools dir when not present. That's it.

tolbertam replied:
changed the error message and removed the else path :+1:

smiklosovic commented on tools/sstableloader/src/org/apache/cassandra/tools/LoaderOptions.java:
```diff
@@ -79,7 +78,6 @@ public class LoaderOptions
      * provide options instead of using these constant fields.
      * @deprecated Use {@code throttle-mib} instead
      */
-    /** @deprecated See CASSANDRA-17677 */
```
@tolbertam are you sure about this? The fact we moved the class from dir to dir does not mean this stopped to be relevant.

tolbertam replied:
I think I dropped it because of the double javadoc.  I think I should just merge it into the doc above, e.g.:

```
@deprecated Use {@code throttle-mib} instead (see CASSANDRA-17677)
```

will make that change.

tolbertam commented on build.xml:
```diff
@@ -2002,18 +2094,26 @@
+  </classpathentry>
   <classpathentry kind="output" path="build/classes/eclipse"/>
   <classpathentry kind="lib" path="test/conf"/>
-  <classpathentry kind="lib" path="${java.home}/../lib/tools.jar"/>
```
tools.jar was removed in [JEP-220](https://openjdk.org/jeps/220) (JDK9): 

> Removed: rt.jar and tools.jar

> The class and resource files previously stored in lib/rt.jar, lib/tools.jar, lib/dt.jar, and various other internal JAR files are now stored in a more efficient format in implementation-specific files in the lib directory. The format of these files is not specified and is subject to change without notice.

Eclipse no longer seems to need this and was failing to build otherwise.

tolbertam commented on build.xml:
```diff
@@ -2025,6 +2125,9 @@
         </classpath>
     </taskdef>
     <mkdir dir=".settings" />
+    <echo file=".settings/org.eclipse.jdt.core.prefs"><![CDATA[eclipse.preferences.version=1
```
Needed because high-scale-lib provides a java.util package which causes eclipses to prevent compilation because the java.util module is detected in the JDK and on the classpath.

![Screenshot From 2025-03-18 00-16-46](https://github.com/user-attachments/assets/d0452186-764c-4407-869c-861e59a8c499)

tolbertam commented on build.xml:
```diff
@@ -2002,18 +2094,26 @@
+  <classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-11">
+      <attributes>
+        <attribute name="module" value="true"/>
+          <attribute name="add-exports" value="java.base/jdk.internal.misc=ALL-UNNAMED:java.base/jdk.internal.ref=ALL-UNNAMED:java.base/jdk.internal.util=ALL-UNNAMED:java.base/sun.nio.ch=ALL-UNNAMED:java.management.rmi/com.sun.jmx.remote.internal.rmi=ALL-UNNAMED:java.rmi/sun.rmi.registry=ALL-UNNAMED:java.rmi/sun.rmi.server=ALL-UNNAMED:java.rmi/sun.rmi.transport.tcp=ALL-UNNAMED:java.sql/java.sql=A ...
```
Just copied these from the definition elsewhere in build.xml.  I'm guessing we could programmatically generate this somehow, but wasn't sure if that was worth the trouble.

tolbertam commented on build.xml:
```diff
@@ -2002,18 +2094,26 @@
   <classpathentry kind="src" output="build/test/fqltool-classes" path="tools/fqltool/test/unit" />
-  <classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER"/>
+  <classpathentry kind="src" output="build/test/sstableloader-classes" path="tools/sstableloader/test/unit" />
+  <classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-11">
```
I opted to configure for Java 11 here as figure it's best to work off the minimal supported version.

tolbertam commented on build.xml:
```diff
@@ -1986,7 +2078,7 @@
     <pathconvert property="eclipse-libs-list" refid="eclipse-project-libs-path" pathsep="${line.separator}">
         <mapper>
-            <regexpmapper from="^(.*)$$" to='&lt;classpathentry kind="lib" path="\1\" \/&gt;'/>
+            <regexpmapper from="^(.*)$$" to='  &lt;classpathentry kind="lib" path="\1" /&gt;'/>
```
This was previously generating invalid xml in the .classpath file:

```
<classpathentry kind="lib" path="/home/tolbertam/Documents/Projects/cassandra/lib/ST4-4.0.8.jar\" \/>
```

In particular the `\/>` part is invalid but I expect `...jar\"` probably would throw things off too.  Also added two whitespaces to the beginning to make the spacing look right (not required, just prettier).

tolbertam commented:
rebased to get in sync with build.xml changes on trunk

tolbertam commented:
Moving this to ready for review now I think I've worked out that last issue with dtests.

michaelsembwever commented on build.xml:
```diff
@@ -664,6 +666,88 @@
+        </fileset>
+      </jar>
+    </target>
+
```
what about putting this to a `tools/sstableloader/build.xml` if we permit cyclic imports or otherwise to a `.build/build-sstableloader.xml`  ,and importing it here ?

This is generally along the lines of what's being done breaking out encapsulated groupings of targets in the .build/*.xml sub-files.  Given this a submodule,  if cyclic imports work (each file imports each other, because these new targets reference `init` and `build`) then it might make sense to put it in that submodule folder rather than under .build/

tolbertam replied:
that sounds good to me, i'll give this a try and see how it goes :+1:

tolbertam replied:
I was skeptical but it looks like circular imports do work!   I'm going to do a little more refactoring to move out sstableloader-y stuff from common tasks and into their own tasks but should have something committed in a little bit.

tolbertam replied:
Just pushed something that I think works pretty well, hopefully that looks ok!

michaelsembwever replied:
looks good.

michaelsembwever commented on .build/sstableloader-deps-template.xml:
i've never settled happy on the naming structure of these new files.

i think i'd prefer `.build/sstableloader-deps-maven-pom.xml` 
as something that was more intuitive.  "template" doesn't tell us much (it's not that important to know upfront that this file gets filtered during the build), but immediate feedback that this is our maven pom dependency declarations is valuable.

but any rename should apply consistently to all such files.

tolbertam replied:
I agree that it definitely is more apparent what the file is for, will make that change.

tolbertam replied:
Changed this, including renaming the other 3 pom files, in 12e681a :+1:

michaelsembwever replied:
looks good.

tolbertam commented on build.xml:
```diff
@@ -360,26 +360,38 @@
-      </fail>
-    </sequential>
-  </macrodef>
+    <macrodef name="create-javadoc">
+      <attribute name="destdir"/>
+      <element name="filesets"/>
+      <sequential>
+        <javadoc destdir="@{destdir}" author="true" version="true" use="true"
+          windowtitle="${ant.project.name} API" classpathref="cassandra.classpath"
+          bottom="Copyright &amp;copy; 2009- The Apache Software Foundation"
+          useexternalfile="yes" encoding="UTF-8" failonerror="false"
+          maxmemory="256m" additionalparam="${jdk11plus-javac-exports}">
+          <filesets/>
+        </javadoc>
+        <fail message="javadoc failed">
+          <condition>
+              <not>
+                  <available file="@{destdir}/index-all.html" />
+              </not>
+          </condition>
+        </fail>
+      </sequential>
+    </macrodef>
```
no changes here other than correcting indentation (can remove if needed)

tolbertam commented on build.xml:
```diff
@@ -360,26 +360,38 @@
+      </sequential>
+    </macrodef>
+
+    <!-- filterset to apply to pom templates -->
+    <filterset id="pom-template">
+        <filter token="version" value="${version}"/>
+        <filter token="final.name" value="${final.name}"/>
+        <filter token="jamm.version" value="${jamm.version}"/>
+        <filter token="allocation-instrumenter.version" value="${allocation-instrumenter.version}"/>
+        <filter token="ecj.version" value="${ecj.version}"/>
+        <filter token="asm.version" value="${asm.version}"/>
+        <filter token="jacoco.version" value="${jacoco.version}"/>
+        <filter token="jflex.version" value="${jflex.version}"/>
+    </filterset>
```
moved from write-poms as is used in sstableloader-write-pom

tolbertam commented on build.xml:
```diff
@@ -2098,6 +2106,16 @@
             file="${build.dir}/${final.name}-javadoc.jar"
             classifier="javadoc"/>
 
+    <!-- the cassandra-sstableloader jar -->
+    <deploy pomFile="${build.dir}/${final.name}-sstableloader.pom"
+            file="${build.dir}/tools/lib/sstableloader.jar"/>
+    <deploy pomFile="${build.dir}/${final.name}-sstableloader.pom"
+            file="${build.dir}/${final.name}-sstableloader-sources.jar"
+            classifier="sources"/>
+    <deploy pomFile="${build.dir}/${final.name}-sstableloader.pom"
+            file="${build.dir}/${final.name}-sstableloader-javadoc.jar"
+            classifier="javadoc"/>
+
```
This is a bit of compromise as I figure `publish` is called explicitly on release, and I figure it likely that one would want to publish all artifacts always.  Alternatively we could use `antcall` and make a `publish-sstableloader` task, but it would redundantly run tasks again.

tolbertam commented on build.xml:
```diff
@@ -1211,10 +1215,12 @@
           <pathelement path="${java.class.path}"/>
           <pathelement location="${stress.build.classes}"/>
           <pathelement location="${fqltool.build.classes}"/>
+          <pathelement location="${sstableloader.build.classes}"/>
```
This is a bit suboptimal, but because `sstableloader-test` makes use of this macro we need to include the classes here as well.  While not perfect it doesn't feel too problematic to have to define these here.

tolbertam commented:
this is going to need a rebasing with the accord merge which i'll look into, but there are some things that accord changed about the build that I really like and I think could motivate further change.  Will spend some time wrapping my head around this and come back to this. Some discussion here: https://the-asf.slack.com/archives/CK23JSY2K/p1744934001097929?thread_ts=1744933008.439629&cid=CK23JSY2K

tolbertam commented:
i'll work on rebasing this this week.

michaelsembwever commented:
`.jenkins/Jenkinsfile` needs editing too.

add the test type here: https://github.com/apache/cassandra/blob/trunk/.jenkins/Jenkinsfile#L166
 (i assume we only need a "small" size)

and add it to all test profiles: https://github.com/apache/cassandra/blob/trunk/.jenkins/Jenkinsfile#L129-L132

tolbertam commented:
Pleasantly surprised at how trivial the merge from trunk after accord was, I put this off a little bit as thought i'd have conflicts everywhere but it was literally just a one-liner:

![image](https://github.com/user-attachments/assets/b758847a-6769-4152-bee6-c0d591c72915)

There was also some changes to `LoaderOptionsTest` that I had to account for where some constants from `TlsTestUtils` were being used that we can't access anymore now that tests are in a separate module.  I figure it'd be fine to just not use the constants since don't think it is worth justifying adding the base test code to sstableloader's test classpath (3d431c0).

Updated the Jenkins file in 87030ca and kicked off a ci test run to ensure everything is still good.

smiklosovic commented:
@tolbertam could you please rebase once more? GitHub tells that it can't be rebased due to conflicts. Rebased on top of the current trunk (and optionally squashed) would be great.

michaelsembwever commented:
> could you please rebase once more? GitHub tells that it can't be rebased due to conflicts. Rebased on top of the current trunk (and optionally squashed) would be great.

Yes, please rebase instead of merge.
(We don't want the `Merge branch 'trunk' into CASSANDRA-20328` commits in this PR)

tolbertam commented:
Sounds good, I figured we'd want to rebase/squash before merge anyways but didn't want to mess with the history in middle of review.  Will take care of that now.

tolbertam committed:
Migrate sstableloader code to its own tools directory and artifact

As part of a broader effort to decouple java driver code from the
server code, this moves sstableloader to its own tools directory.

As sstableloader is also used as a library (CASSANDRA-10637), added
a new artifact 'cassandra-sstableloader' that will get deployed to
maven along with 'cassandra-all'.

While I expect this is likely a niche use case, this will allow users
to continue using BulkExport as a library.

Moves sstableloader-specific targets to its own build.xml in
tools/sstableloader/build.xml.

Also updates IDE project files and circleci to utilize new
sstableloader-specific targets.

patch by Andy Tolbert; reviewed by Stefan Miklosovic and Mick Semb Wever for CASSANDRA-20328

tolbertam commented:
squashed into 1 commit, note that two tiny changes included in the commit that weren't there before:

* Added a CHANGES.txt entry
* Removed an extraneous space I caught in 'exit 2' at line 42 in bin/sstableloader.

Other than this it is even with the original series of commits (87030caf) + commits since I merged.

tolbertam commented:
I found a few more issues with the circle configuration, generated the paid configuration on a separate branch and ran the sstableloader specific jobs and seems to be working rather well:

https://app.circleci.com/pipelines/github/tolbertam/cassandra/182/workflows/a51cdc0c-596a-4fb6-91d3-0afd5a3b4165

Hopefully I've addressed everything here.  I'm doing some more sanity testing, if anything else pops up let me know.  When everyone is comfortable with where things are at I can do one last squash :+1:

tolbertam commented:
One thing I realized is that for the python-dtests to pass, we'll need to update the cassandra-test tag in ccm: (eg https://app.circleci.com/pipelines/github/tolbertam/cassandra/182/workflows/a51cdc0c-596a-4fb6-91d3-0afd5a3b4165/jobs/15909/tests). ok if we do that?

smiklosovic reviewed (approved):
LGTM minus https://github.com/apache/cassandra/pull/3899#issuecomment-2878325825

tolbertam commented:
squashed with no changes before a final round of testing.

michaelsembwever commented:
running ci: http://35.228.140.106:8080/job/cassandra/238/

EDIT/UPDATE: results here: https://issues.apache.org/jira/browse/CASSANDRA-20328?focusedCommentId=17952433&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-17952433

tolbertam commented:
Thanks for running tests and reviewing!  I'll fix up the changelog and merge this shortly.

tolbertam commented:
Committed as f80a234d1ec33dbe927bc8af28f1a290335819a3, thanks all!
